Implement new UI for switching languages

jsnajdr opened the pull request:
This is a PR for my trial project described in my post on the i18n P2 blog (internal ref: pxLjZ-3EJ-p2).

The main part of the effort is implementing a new language picker component (internal ref: p4TIVU-5np-p2).

A big task with many subtasks:
- [x] Implement a `LanguageChooser` component intended to replace the existing `LanguageSelector`. Its design is based on the `SitesDropdown` component.
- [x] Implement a `LanguageChooserModal` dialog that opens after clicking the "Change" link in the `LanguageChooser`. The design is based on `EditorMediaModal`. A tab bar with language groups and search, filtered list of languages to choose from, and  a button bar with "Cancel" and "Select" buttons.
- [x] Use the new `LanguageChooser` on the `/me/account` page.
- [x] Use the `LanguageChooser` on the `my-sites/site-settings/form-general` page. This will need several improvements to the component:
  - [x] In addition to `valueLink` prop, support also a `value`/`onChange` variant. Or get rid of the `valueLink` in `me/account/main.jsx`. That might be more future-proof, as `valueLink` is deprecated. (`valueLink` is being removed in PR #12729)
  - [x] Support the `disabled` state
  - [x] Support hooking telemetry (`eventTracker`) into the component
- [x] Support a "placeholder" state in `LanguageChooser`. It's displayed when settings data is being loaded.
- [x] Highlight the currently selected language in `LanguageChooserModal`
- [x] Add a `README.md` for the new components
- [x] Learn how unit tests are written in Calypso and write a testing suite (are we testing actions and reducers, or also UI rendering? What about integration (e2e) tests?
- [x] ~~Display the explanation text side-by-side with the `LanguageChooser` (if the viewport is wide enough)~~ abandoned, displaying under the component for consistency.
- [x] Create a `LanguageSidebarChooser` component that will be displayed in sidebars
- [x] Add a "Suggested Languages" list to `LanguageChooserModal`. The suggestions will be based on the browser language settings.
  - [x] Get the language list from `navigator.languages`
  - [x] ~~For browsers that don't support `navigator.languages` (it's a new experimental API, supported only by Chrome and Firefox atm, according to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorLanguage/languages), find some fallback - probably some server REST API endpoint that reads the `Accept-Language` header.~~ Decided to support only `navigator.languages` now and maybe make `/locale-guesser` smarter sometime in future.
  - [x] As a last-resort fallback, use geolocation for suggestions. Investigate the `/locale-guesser` API endpoint. Result: using geolocation to select the default tab.
- [x] Group the language list into territories (Europe, Asia, Americas, ...), like the Facebook language picker does
  - [x] Use the CLDR data for that. Decide whether to hardcode the data into the `config/_shared.json` file, or whether to query it online (see below)
  - [x] Investigate the [wp-cldr](https://github.com/Automattic/wp-cldr) plugin. Does it have REST API? Can we detect if it's installed and query it? (It should be available on all WordPress.com sites). Result: hardcoding the data into JS code (it's not that big). Might load the data with `AsyncLoad` in future.
- [ ] Support searching for a language by multiple names (Czech, Čeština, Tschechisch, ...). CLDR has all the translations.
  - [ ] How big will the translation data be? Decide whether to support only popular languages, or all.
  - [ ] Use the same data to display a tooltip over the language name that translates the name into the current language. The language name in the list is always self-referential (English, Deutsch, Čeština). Facebook dialog does this.
- [x] Tune the CSS styling of the modal. Right now, the styles depend on precise pixel sizes. Vertical flex should make the styling simpler and more reliable. Do all supported browser have good-enough vertical flex support?
- [x] Fine-tune the responsivity of the modal dialog (size, column sizes and count)
- [x] ~~When a language name is too long and doesn't fit into the column, add a "fade-out" effect to make the clipping look good.~~ (turns out this is not needed when the column count is responsive -- there's always enough room for all existing language names)
- [x] Clean up the language list data in `config/_shared.json`. There are a few errors. (moved this task to an independent PR: #12714)
  - [x] Don't include the language slug in the language `name`. Convert `en - English` to `English`.

A related task that will probably be an independent PR: avoid reloading the whole page after a language is switched. Just trigger a React rerender with the new language. (Moved to #13712)
- [ ] Does the `localize` HOC from `i18n-calypso` support this? Will the value of the injected `translate` property change?
- [ ] Will all components properly rerender when `translate` props changes? Or will it be blocked by some `shouldComponentUpdate` implementation? Or a wrong/missing `componentWillReceiveProps`?
- [ ] Does something that's not managed by React need to change on language switch? Like CSS styles or other assets?
- [x] Decide what directory should the new components be in. Is it `client/components`, or `client/blocks` or `client/components/form`? **Result:** It should be `client/component` - it's a "dumb" component that only renders what it's passed in props. To be eligible to be a `block`, it would need to connect to the app state by itself.
- [x] Consider removing the current `LanguageSelector` component and replace it completely with the `LanguageChooser`, maybe even reusing the name.

The `LanguageChooser` uses the `lib/user-settings` module that uses a legacy, event-based approach to data. Converting this module to Redux would be great.
- [x] Create a Redux state slice for user settings and provide appropriate action creators and selectors as the new API (happens in a separate PR: #12819)
- [ ] Convert the existing `lib/user-settings` into a thin wrapper over the Redux API
- [ ] Migrate all usages of `lib/user-settings`
- [ ] Remove the legacy `lib/user-settings` module

jsnajdr commented:
> I have tried it out on my machine: great progress! (Please tell us when you're ready for UI feedback :) ) Does the list represent the priority sequence in which you plan to implement the other points?

The list is in a rather random order - it's a memory dump where I wanted to write things down before I forget them. I believe that all or almost all of them need to be implemented before the feature is shippable. Is there anything that you think is likely to spark some longer discussion and that should be implemented sooner rather than later?

> It would be great if you could provide screenshots of the current implementation state, it's both great for documenting the progress as well as it allows others to quickly peak at the current stage.

Here is a screenshot of the state as of today:

![language-chooser-03-27](https://cloud.githubusercontent.com/assets/664258/24375276/0b93cca6-1338-11e7-8d9a-b21599e84561.png)

It shows a part of the `my/account` form with the `LanguageChooser` and also the `SitesDropdown` for comparison. On the right, there is the `LanguageChooserModal` dialog.

A design question for @folletto: when I see the `LanguageChooser` and `SitesDropdown` next to each other, I feel they don't fit together very well. The "English" and "Change" labels are big and bold, while the text in the `SitesDropdown` looks more subtle. I tried to change the font weight and sizes slightly, and the result is in the bottom left. I like it more. What do you think?

folletto commented:
> I tried to change the font weight and sizes slightly, and the result is in the bottom left. I like it more. What do you think?

A step back for the rationale: it was bold to match the button styling, not the site dropdown styling. 

I agree there's a mismatch there tho, as the site dropdown doesn't adjust weight to align with buttons, so it's probably something to be reviewed more broadly as there's an existing incoherence there.

Given we're "in between", I'm ok with either option. If you feel the normal weight one is better, go for it. :)

ehg commented on client/components/language-chooser/index.jsx:
```diff
@@ -0,0 +1,108 @@
+		} ),
+	}
+
+	constructor( props ) {
```
Looks like you could use ESNext property initialisers instead of this constructor.

jsnajdr replied:
Fixed.

ehg commented on client/components/language-chooser/language-chooser-modal.jsx:
```diff
@@ -0,0 +1,127 @@
+import React, { PureComponent } from 'react';
```
Missing an `External dependencies` docblock

jsnajdr replied:
Fixed.

a8ci18n commented on client/components/language-chooser/language-chooser-modal.jsx:
```diff
@@ -0,0 +1,179 @@
+			},
+			{
+				action: 'confirm',
+				label: translate( 'Select Language' ),
```
Hi! I've found a possible matching string that has already been translated **60** times:
`translate( 'Select a language below' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jsnajdr commented:
Today's update:
- added a language button into the sidebar footer
- cleaned up the language chooser modal:
  - highlight the selected language in the dialog
  - save and close the dialog only after clicking the confirmation button

<img width="723" alt="sidebar-button-and-selection" src="https://cloud.githubusercontent.com/assets/664258/24428312/51075ae2-140e-11e7-9227-dedad35e4d38.png">

New problem: in order to be able to react to changes in the user settings, and to do that without using legacy components like the `observe` mixin, we'll need to move the `user-settings` module functionality (fetching, getting, updating, saving) to a Redux state subtree with appropriate actions, reducers and selectors. Quite a big task.

As a rather ugly temporary workaround, I now subscribe to `userSettings`' `change` event in `LanguageButton`'s `componentDidMount`/`componentWillUnmount`.

jsnajdr commented:
Added a list of suggested languages:

<img width="764" alt="suggested-languages" src="https://cloud.githubusercontent.com/assets/664258/24518578/bfe3f888-1582-11e7-911e-d5f561601991.png">

It's based on browser preferences, i.e., on `navigator.languages`. If the browser doesn't support this feature, the suggested list is not displayed at all.

What remains to be done on this feature:
- [ ] make it responsive - in a narrow viewport, the list overflows and is not pretty.
- [ ] make the selection highlight box the same size in both the language list and the selected language list. Now the heights are different, for some strange reasons.
- [ ] test on Safari (Calypso in devel mode often crashes it)
- [ ] test on IE

@folletto What do you think are best options for making the suggested language list responsive? Show limited number of suggestions on one row and hide the ones that would overflow? Convert the list into a popup menu when the viewport width is small? (`SectionNav` does exactly this with tabs, maybe we can reuse it)

deBhal commented on client/components/language-chooser/index.jsx:
```diff
@@ -0,0 +1,103 @@
+					</div>
+				</div>
+				<LanguageChooserModal
+					isVisible={ this.state.open }
```
Does the LanguageChooserModal do anything if it's not visible or own the visibility in some sense? If not, we might take the isVisible out of it and check it here to simplify the component and avoid creating it at all when it's not visible.

jsnajdr commented:
Latest update: while the user settings are being fetched, all the language chooser buttons show a placeholder with the 'loading' status:

![settings-placeholder](https://cloud.githubusercontent.com/assets/664258/24599998/08662a90-1853-11e7-85b4-a2d2eafefee4.gif)

My next steps will be breaking this big PR into smaller ones that can be independently reviewed and merged.

jsnajdr commented:
Rebased this PR to latest master, where #12714 just got merged.

jsnajdr commented:
Current status:
- rebased the whole PR on top of #12729 (removal of `valueLink` usages) and changed the API of `LanguageChooser` to no longer support `valueLink`
- used the `LanguageChooser` on the Site Settings form

@akirk @folletto @ehg I guess I'm ready for a design review and code review of the new language picker. There are currently 5 commits in this branch that I indend to merge as separate PRs.

The first version of the chooser dialog will have two tabs with "Popular Languages" and "All Languages", there will be no suggested languages and no division by territory. These will be followups after the basic version is accepted and merged.

I'll be now working on Reduxization of the `lib/user-settings` module. Although it will be quite a big change, it won't affect the language picker code very much.

folletto commented:
> There are currently 5 commits in this branch that I indend to merge as separate PRs.

Ok go ahead for the separate PRs.

In general the design looks good.

I've just a secondary note that the "button" should be entirely clickable, and should highlight similar to the standard buttons on hover, not just "Change". Sorry if it wasn't clear from the mockups.

jsnajdr commented:
> I've just a secondary note that the "button" should be entirely clickable, and should highlight similar to the standard buttons on hover, not just "Change". Sorry if it wasn't clear from the mockups.

Oh, it was the "Change" link that made me think that it's the only active part of the button. I changed the following:
- make the whole button clickable
- add a border highlight on hover - exactly the same one as appears on `<input>` form elements
- The "Change" label is no longer an `<a>` element - we don't want it to use link styles like a diferent color on hover. It's just a `<span>` with text color identical to a `button.is-compact`

I noticed that the `<SitesDropdown>` button also doesn't have the hover highlight. I'll address that in another PR.

folletto commented:
Ok, looks good:

![cap-](https://cloud.githubusercontent.com/assets/4389/24799592/f27b4cb2-1b93-11e7-9875-8d66bd3960b2.gif)

a8ci18n commented on client/components/language-picker/modal.jsx:
```diff
@@ -0,0 +1,256 @@
+			},
+			{
+				action: 'confirm',
+				label: translate( 'Select Language' ),
```
Hi! I've found a possible matching string that has already been translated **60** times:
`translate( 'Select a language below' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/language-chooser)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

yoavf commented:
Suggestion: use the [Autonym Font](https://github.com/santhoshtr/AutonymFont) for the language names to prevent ܕܥܒܪܸܝܛ 

<img width="600" alt="screen shot 2017-04-20 at 11 30 12" src="https://cloud.githubusercontent.com/assets/844866/25221157/c25391a6-25bc-11e7-8870-4769a4a915da.png">

jsnajdr commented:
> Suggestion: use the Autonym Font for the language names to prevent ܕܥܒܪܸܝܛ

Looks like a great idea, thanks! I'll definitely try the font out. I'm wondering how the font glyph look like design-wise. Also, I'm a bit worried that the project is explicitly marked as unmaintained.

@yoavf do you have any direct experience with the font? Have used it on any project?

jsnajdr commented:
Update: using the new Redux `state/user-settings` module to implement the sidebar button. Works very well, except there are some problems reacting to success/error response after saving the new language. Will blog about it.

yoavf commented:
>  Also, I'm a bit worried that the project is explicitly marked as unmaintained.
@yoavf do you have any direct experience with the font? Have used it on any project?

Nope, no personal experience. I was thinking it was used on Wikipedia, but I've just checked and this doesn't seem to be the case. Feel free to ignore it then :)

jsnajdr commented:
Closing this PR as all related changes were moved to separate smaller PRs under the [Language Picker Project](https://github.com/Automattic/wp-calypso/projects/31)
